### PR TITLE
Devices actions

### PR DESCRIPTION
**Pull request by chrisnojima**
Cleanup of devices actions

Key ideas:
- Mostly pure
- Each function does one simple thing, takes an action/state or some data and outputs an action
- Multiple listeners for actions
- Action listener handles one type of data. So we have 1 function `changeWaiting` that's in charge of the waiting state, and not multiple actions which might be mutating it
- Using opaque types for DeviceID (not supported by our version of prettier but newer version does work. Short term had to add types/devices to eslintignore :( (we should use this in chat and elsewhere we have multiple id types)

### CODE REVIEW

**Comment by chrisnojima** [comment]
@keybase/react-hackers please take a look and give comments here or on chat

**Comment by chrisnojima** [comment]
(i didn't implement the commented out stuff because i wanted to discuss this first half first)

**Comment by MarcoPolo** [comment]
So in this case, the reducer is nothing more than a simple key-value db, is that correct?

**Comment by chrisnojima** [comment]
Basically. It can be hierarchical but hopefully not

**Comment by chrisnojima** [comment]
- [ ] rebase against master

**Comment by chrisnojima** [comment]
moving forward with this change

File: shared/actions/__tests__/devices.test.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+      DevicesGen.createPaperKeyCreated({paperKey: new HiddenString('')}),
+      DevicesGen.createDeviceRevoked({
+        deviceID: '',
+        deviceName: '',
+        wasCurrentDevice: false,
+      }),
+    ]
+
+    actionsThatDecrement.forEach(a =>
+      expect(_testing.changeWaiting(a)).toEqual(Saga.put(WaitingGen.createDecrementWaiting(payload)))
```
Comment (MarcoPolo) [inline shared/actions/__tests__/devices.test.js]: Nice, it's pretty simple 👍

**Comment by chrisnojima** [comment]
- [ ] opaque deviceid

File: .eslintignore [hunk 2]
```diff
@@ -18,5 +18,6 @@
 /shared/markdown/parser.js
 /shared/node_modules
 /shared/test/setupTestFrameworkScriptFile.js
+/shared/constants/types/devices.js
```
Comment (chrisnojima) [inline .eslintignore]: TODO remove when prettier gets updated this week

File: shared/actions/devices.js [hunk 3]
```diff
@@ -1,188 +1,224 @@
...
-  yield Saga.safeTakeLatest(DevicesGen.load, _deviceListSaga)
-  yield Saga.safeTakeEvery(DevicesGen.revoke, _deviceRevokedSaga)
-  yield Saga.safeTakeEvery(DevicesGen.paperKeyMake, _devicePaperKeySaga)
-  yield Saga.safeTakeEvery(DevicesGen.showRevokePage, _deviceShowRevokePageSaga)
+  // Load devices
+  yield Saga.safeTakeLatestPure(
+    DevicesGen.devicesLoad,
+    rpcDeviceList,
+    dispatchDevicesLoaded,
+    dispatchDevicesLoadedError
```
Comment (chrisnojima) [inline shared/actions/devices.js]: since this is a take latest we can get canceled.

File: shared/logger/index.js [hunk 4]
```diff
@@ -27,11 +27,13 @@ function _mergeSortedArraysHelper<A>(sortFn: (a: A, b: A) => number, ...arrays:
...
-      return unlink(filepath)
-    }
-  })
+  return stat(filepath)
+    .then(({lastModified}) => {
+      if (Date.now() - lastModified > olderThanMs) {
+        return unlink(filepath)
+      }
+    })
+    .catch(() => {})
```
Comment (chrisnojima) [inline shared/logger/index.js]: else if the file isn't there we get warnings

File: shared/constants/types/devices.js [hunk 5]
```diff
@@ -16,7 +17,28 @@ export type _DeviceDetail = {
...
 export type _State = {
-  deviceIDs: I.List<string>,
-  waitingForServer: boolean,
+  idToDetail: I.Map<DeviceID, DeviceDetail>,
+  idToEndangeredTLFs: I.Map<DeviceID, I.Set<string>>,
 }
 export type State = I.RecordOf<_State>
+
+// Converts a string to the DeviceType enum, logging an error if it doesn't match
+export function stringToDeviceType(s: string): DeviceType {
```
Comment (chrisnojima) [inline shared/constants/types/devices.js]: type conversions should be here so the opaque typing works

File: shared/devices/container.js [hunk 6]
```diff
@@ -3,85 +3,94 @@ import Devices from '.'
...
+  connect(mapStateToProps, mapDispatchToProps, mergeProps),
   lifecycle({
-    componentWillMount: function() {
-      this.props.loadDevices()
+    componentWillMount() {
+      this.props._loadDevices()
     },
   }),
-  // Don't pass immutable things to dumb components
-  mapProps(props => ({
```
Comment (chrisnojima) [inline shared/devices/container.js]: with mergeProps we don't need mapProps usually

File: shared/devices/device-page/index.desktop.js [hunk 7]
```diff
@@ -62,41 +62,32 @@ const Timeline = ({timeline}) => (
...
-  revokedAt,
-  showRevokeDevicePage,
-  onBack,
-  bannerBackgroundColor,
-  bannerColor,
-  bannerDesc,
-  icon,
-  revokeName,
-}: Props) => (
+const Render = (props: Props) => (
```
Comment (chrisnojima) [inline shared/devices/device-page/index.desktop.js]: this top level type spread doesn't really give us much, just adds a bunch of lines

File: shared/util/saga.js [hunk 8]
```diff
@@ -159,15 +159,15 @@ function safeTakeEveryPure<A, R, FinalAction, FinalActionError>(
...
+  actionCreatorsWithError?: (result: R, action: A) => FinalActionError
 ) {
-  return safeTakeLatest(pattern, function* safeTakeLatestPureWorker(action: A) {
+  const safeTakeLatestPureWorker = function* safeTakeLatestPureWorker(action: A) {
     // If the pureWorker fn takes two arguments, let's pass the state
     try {
       let result
       if (pureWorker.length === 2) {
-        const state: TypedState = yield select()
+        const state: TypedState = yield select(s => s)
```
Comment (chrisnojima) [inline shared/util/saga.js]: flow got very confused with the select() syntax here, so passing s=>s fixes that

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is ready

File: shared/logger/index.js (hunk 4 above)
Comment (buoyad) [reply shared/logger/index.js]: I noticed this earlier too, should we create the file if it's not there?

File: shared/constants/types/devices.js [hunk 9]
```diff
@@ -16,7 +17,28 @@ export type _DeviceDetail = {
...
+    case 'desktop':
+    case 'backup':
+      return s
+    default:
+      console.log('Unknown Device Type %s. Defaulting to `desktop`', s)
+      return 'desktop'
+  }
+}
+
+export function stringToDeviceID(s: string): DeviceID {
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/constants/types/devices.js]: I don't think we really want to do this. As far as the front end is concerned this is an opaque type given to us. We don't / shouldn't have a way to validate these ids. we just have to make sure we handle them internally correctly

**Pull request merged**